src/faucet/server_api.rs

```diff
@@ -189,3 +192,158 @@ pub async fn signed_erc20_transfer(
+}
+
+#[cfg(feature = "ssr")]
+fn parse_and_validate_address(
+    address: &str,
+    network: fvm_shared::address::Network,
+    response: &leptos_axum::ResponseOptions,
+) -> Result<fvm_shared::address::Address, ServerFnError> {
+    match crate::utils::address::parse_address(address, network) {
+        Ok(addr) => Ok(addr),
+        Err(e) => {
+            log::error!("Invalid address: {}", e);
+            response.set_status(StatusCode::BAD_REQUEST);
+            Err(ServerFnError::ServerError(format!(
+                "Invalid address: {}",
+                e
+            )))
+        }
+    }
+}
+
+#[cfg(feature = "ssr")]
+async fn handle_native_claim(
+    faucet_info: FaucetInfo,
+    address: String,
+    response: &leptos_axum::ResponseOptions,
+) -> Result<String, ServerFnError> {
+    use crate::utils::message::message_transfer;
+    use crate::utils::rpc_context::Provider;
+    use fvm_shared::address::set_current_network;
+
+    let network = faucet_info.network();
+    set_current_network(network);
+
+    let recipient = parse_and_validate_address(&address, network, response)?;
+    let rpc = Provider::from_network(network);
+
+    let id_address = rpc.lookup_id(recipient).await.unwrap_or_else(|_| {
+        log::debug!("ID lookup failed, using recipient address: {:?}", recipient);
+        recipient
+    });
+    let from = faucet_address(faucet_info)
+        .await?
+        .to_filecoin_address(network)
+        .map_err(ServerFnError::new)?;
+    let nonce = rpc
+        .mpool_get_nonce(from)
+        .await
+        .map_err(ServerFnError::new)?;
+    let drip_amount = faucet_info.drip_amount();
+    let raw_msg = message_transfer(from, id_address, drip_amount.clone());
+    let msg = rpc
+        .estimate_gas(raw_msg)
+        .await
+        .map_err(ServerFnError::new)?;
+
+    match signed_fil_transfer(
+        LotusJson(id_address),
+        msg.gas_limit,
+        LotusJson(msg.gas_fee_cap),
+        LotusJson(msg.gas_premium),
+        nonce,
+        faucet_info,
+    )
+    .await
+    {
+        Ok(LotusJson(smsg)) => {
+            let cid = rpc.mpool_push(smsg).await.map_err(ServerFnError::new)?;
+            Ok(cid.to_string())
+        }
+        Err(err) => Err(handle_faucet_error(response, err)),
+    }
+}
+
+#[cfg(feature = "ssr")]
+async fn handle_erc20_claim(
+    faucet_info: FaucetInfo,
+    address: String,
+    response: &leptos_axum::ResponseOptions,
+) -> Result<String, ServerFnError> {
+    use crate::utils::address::AddressAlloyExt;
+    use crate::utils::rpc_context::Provider;
+    use fvm_shared::address::set_current_network;
+
+    let network = faucet_info.network();
+    set_current_network(network);
+
+    let recipient = parse_and_validate_address(&address, network, response)?;
+    let rpc = Provider::from_network(network);
+
+    let owner_fil_address = faucet_address(faucet_info)
+        .await?
+        .to_filecoin_address(network)
+        .map_err(ServerFnError::new)?;
```

Comment:
You can place this part directly into the `claim_token` as it is being used in both method (handle_erc20_claim, handle_native_claim), and pass the necessary values as the method param.